presto-raptor/src/main/java/com/facebook/presto/raptor/storage/InplaceShardRewriter.java

```diff
@@ -0,0 +1,136 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptor.storage;
+
+import com.facebook.airlift.json.JsonCodec;
+import com.facebook.presto.raptor.backup.BackupManager;
+import com.facebook.presto.raptor.metadata.ShardDelta;
+import com.facebook.presto.raptor.metadata.ShardInfo;
+import com.facebook.presto.raptor.metadata.ShardRecorder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static com.facebook.airlift.concurrent.MoreFutures.getFutureValue;
+import static com.facebook.airlift.json.JsonCodec.jsonCodec;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+
+public class InplaceShardRewriter
+        implements ShardRewriter
+{
+    private static final JsonCodec<ShardDelta> SHARD_DELTA_CODEC = jsonCodec(ShardDelta.class);
+    private final OrcStorageManager orcStorageManager;
+    private final StorageService storageService;
+    private final ShardRecorder shardRecorder;
+    private final BackupManager backupManager;
+    private final FileSystem fileSystem;
+    private final long transactionId;
+    private final OptionalInt bucketNumber;
+    private final UUID shardUuid;
+    private final Map<String, Type> columns;
+    private final ExecutorService deletionExecutor;
+    private final String nodeId;
+
+    public InplaceShardRewriter(
+            FileSystem fileSystem,
+            long transactionId,
+            OptionalInt bucketNumber,
+            UUID shardUuid,
+            Map<String, Type> columns,
+            ExecutorService deletionExecutor,
+            String nodeId,
+            OrcStorageManager orcStorageManager,
+            StorageService storageService,
+            ShardRecorder shardRecorder,
+            BackupManager backupManager)
+    {
+        this.orcStorageManager = orcStorageManager;
```

Comment:
Compared to `DeltaShardRewriter`, `InplaceShardRewriter` seems to be more coupled with `OrcStorageManager` as it's originally part of `OrcStorageManager`. I suggest we can keep `InplaceShardRewriter` an embedded class of `OrcStorageManager`or an anonymous class.

Reply:
@highker what do you think of this suggestion?